### ISSUE TITLE
Proactive Intelligence: Multi-Source Context Beyond Email

### ISSUE THREAD

**Issue by ckorhonen** [ckorhonen/openpoke#4]
## Overview
Expand OpenPoke's context awareness beyond email to create a truly proactive intelligence system that aggregates signals from multiple data sources, enabling more accurate and timely insights.

## 🎯 Objectives
- Build a scalable ML pipeline for multi-source data ingestion and processing
- Integrate diverse data sources (calendar, tasks, documents, chat, etc.)
- Develop unified context representation and feature extraction
- Create intelligent priority scoring across all data sources
- Enable proactive suggestions based on holistic context

### High-Level Architecture
```
┌─────────────────────────────────────────────────────────────────┐
│                    Data Source Layer                             │
├──────────────┬──────────────┬──────────────┬──────────────────┤
│   Email      │   Calendar   │   Tasks      │   Documents      │
│   (IMAP)     │   (CalDAV)   │   (REST)     │   (File/Cloud)   │
└──────┬───────┴──────┬───────┴──────┬───────┴──────┬───────────┘
       │              │              │              │
       v              v              v              v
┌─────────────────────────────────────────────────────────────────┐
│                  Data Ingestion Layer                            │
│  - Source-specific connectors                                    │
│  - Rate limiting & backoff                                       │
│  - Delta sync & incremental updates                              │
│  - Error handling & retry logic                                  │
└────────────────────────────┬────────────────────────────────────┘
                             v
┌─────────────────────────────────────────────────────────────────┐
│                  Normalization Layer                             │
│  - Schema standardization                                        │
│  - Entity extraction (people, dates, locations)                  │
│  - Content cleaning & preprocessing                              │
│  - Metadata enrichment                                           │
└────────────────────────────┬────────────────────────────────────┘
                             v
┌─────────────────────────────────────────────────────────────────┐
│                  Feature Engineering Layer                       │
│  - Text embeddings (sentence transformers)                       │
│  - Temporal features (urgency, deadlines)                        │
│  - Relationship graphs (sender/attendee networks)                │
│  - Historical interaction patterns                               │
└────────────────────────────┬────────────────────────────────────┘
                             v
┌─────────────────────────────────────────────────────────────────┐
│                  ML Models & Scoring                             │
│  - Priority scoring model                                        │
│  - Topic classification                                          │
│  - Sentiment analysis                                            │
│  - Action item detection                                         │
│  - Context similarity matching                                   │
└────────────────────────────┬────────────────────────────────────┘
                             v
┌─────────────────────────────────────────────────────────────────┐
│                  Context Aggregation Layer                       │
│  - Cross-source entity resolution                                │
│  - Context fusion & deduplication                                │
│  - Temporal context windows                                      │
│  - Relevance ranking                                             │
└────────────────────────────┬────────────────────────────────────┘
                             v
┌─────────────────────────────────────────────────────────────────┐
│                  Intelligence Layer                              │
│  - Proactive suggestion generation                               │
│  - Priority notifications                                        │
│  - Context-aware summaries                                       │
│  - Intelligent scheduling & reminders                            │
└─────────────────────────────────────────────────────────────────┘
```

### Phase 1: Core Sources
1. **Email (Enhanced)**
   - Current IMAP integration
   - Add thread grouping
   - Attachment analysis
   - Sender reputation scoring

2. **Calendar**
   - CalDAV/iCal integration
   - Event extraction: meetings, deadlines, appointments
   - Conflict detection
   - Pre-meeting context aggregation
   - Features:
     - Attendee lists
     - Location data
     - Event descriptions
     - Recurrence patterns

3. **Task Management**
   - REST API integrations: Todoist, Asana, Trello, Linear
   - Task attributes: status, priority, assignee, due date
   - Project/board context
   - Dependency tracking

### Phase 2: Extended Sources
4. **Documents & Files**
   - Google Drive, Dropbox, OneDrive integration
   - Recent activity tracking
   - Shared document detection
   - Version change notifications

5. **Chat & Messaging**
   - Slack, Discord, Teams integrations
   - DM priority detection
   - @mention extraction
   - Channel activity summaries

6. **Browser Activity**
   - Open tabs context
   - Recent bookmarks
   - Search history (opt-in)
   - Active project detection

### Phase 3: Advanced Sources
7. **Code Repositories**
   - GitHub/GitLab integration
   - PR review requests
   - Issue assignments
   - Commit activity

8. **CRM & Sales**
   - Salesforce, HubSpot integration
   - Deal stage tracking
   - Customer interaction timeline

### Data Models

```python
# Unified Event Schema
class UnifiedEvent:
    id: str
    source: DataSource  # EMAIL, CALENDAR, TASK, etc.
    source_id: str
    timestamp: datetime
    event_type: EventType
    content: str
    entities: List[Entity]  # people, orgs, locations, dates
    metadata: Dict[str, Any]
    embedding: np.ndarray
    priority_score: float
    urgency_score: float
    context_tags: List[str]
    related_events: List[str]  # IDs of related events

# Entity Extraction
class Entity:
    type: EntityType  # PERSON, ORG, DATE, LOCATION, AMOUNT
    value: str
    canonical_id: str  # for entity resolution
    confidence: float
    source_span: Tuple[int, int]

# Context Window
class ContextWindow:
    focal_event: UnifiedEvent
    related_events: List[UnifiedEvent]
    time_range: Tuple[datetime, datetime]
    relevance_scores: List[float]
    summary: str
    action_items: List[ActionItem]
```

#### 1. Priority Scoring Model
```python
class PriorityScorer:
    """Multi-factor priority scoring"""
    
    def __init__(self):
        self.model = GradientBoostingRegressor()
        self.feature_extractors = [
            TemporalUrgencyExtractor(),
            SenderImportanceExtractor(),
            ContentComplexityExtractor(),
            HistoricalPatternExtractor(),
            CrossSourceRelevanceExtractor()
        ]
    
    def score(self, event: UnifiedEvent, context: ContextWindow) -> float:
        features = self._extract_features(event, context)
        return self.model.predict(features)[0]
    
    def _extract_features(self, event, context):
        return np.concatenate([
            extractor.extract(event, context)
            for extractor in self.feature_extractors
        ])
```

#### 2. Context Similarity Engine
```python
class ContextSimilarity:
    """Find related events across sources"""
    
    def __init__(self):
        self.encoder = SentenceTransformer('all-MiniLM-L6-v2')
        self.index = faiss.IndexFlatIP(384)  # dimension of embeddings
    
    def find_related(self, event: UnifiedEvent, k: int = 10) -> List[UnifiedEvent]:
        query_embedding = event.embedding
        distances, indices = self.index.search(query_embedding, k)
        return self._fetch_events(indices)
```

#### 3. Action Item Detector
```python
class ActionItemDetector:
    """Extract actionable items from content"""
    
    def __init__(self):
        self.model = pipeline("token-classification", 
                             model="dslim/bert-base-NER")
        self.patterns = [
            r"(?i)(please|could you|can you|need to|must|should)\s+(.+?)(?=\.|$)",
            r"(?i)(todo|to-do|action item|ai):\s*(.+?)(?=\n|$)",
            r"(?i)(deadline|due|by)\s+(\d{1,2}/\d{1,2}|\w+\s+\d{1,2})"
        ]
    
    def detect(self, content: str) -> List[ActionItem]:
        # Combine NER + pattern matching + dependency parsing
        pass
```

### Data Ingestion Pipeline

```python
class DataIngestionPipeline:
    """Orchestrates data ingestion from all sources"""
    
    def __init__(self):
        self.connectors = {
            DataSource.EMAIL: EmailConnector(),
            DataSource.CALENDAR: CalendarConnector(),
            DataSource.TASKS: TaskConnector(),
            # ... more connectors
        }
        self.normalizer = EventNormalizer()
        self.feature_extractor = FeatureExtractor()
        self.storage = EventStore()
    
    async def sync_all(self):
        """Parallel sync of all data sources"""
        tasks = [
            self._sync_source(source)
            for source in self.connectors.keys()
        ]
        results = await asyncio.gather(*tasks, return_exceptions=True)
        return self._handle_results(results)
    
    async def _sync_source(self, source: DataSource):
        connector = self.connectors[source]
        raw_events = await connector.fetch_updates()
        
        for raw_event in raw_events:
            # Normalize
            unified_event = self.normalizer.normalize(raw_event, source)
            
            # Extract features
            unified_event = self.feature_extractor.process(unified_event)
            
            # Store
            await self.storage.save(unified_event)
```

### Week 1-2: Foundation & Architecture
**Goals:** Set up core infrastructure and data models

- [ ] Design and implement unified event schema
- [ ] Set up data ingestion framework (async workers, queue system)
- [ ] Implement base connector interface
- [ ] Set up feature store (vector DB + metadata DB)
- [ ] Create data normalization pipeline
- [ ] Set up monitoring and logging infrastructure

**Deliverables:**
- Data models documented and implemented
- Base ingestion framework operational
- Development environment configured

### Week 3-4: Core Integrations
**Goals:** Implement Phase 1 data sources

- [ ] Enhanced email connector with thread analysis
- [ ] Calendar integration (CalDAV)
  - Google Calendar
  - iCal support
- [ ] Task management integration (start with 1-2 APIs)
  - Todoist integration
  - Linear integration
- [ ] Entity extraction pipeline
  - People, dates, locations
  - Entity resolution across sources
- [ ] Basic priority scoring model (v1)

**Deliverables:**
- 3+ data sources actively ingesting
- Basic entity extraction working
- Initial priority scores generated

### Week 5: ML Pipeline & Features
**Goals:** Build core ML capabilities

- [ ] Implement text embedding pipeline
  - Choose and deploy sentence transformer model
  - Batch embedding generation
- [ ] Build feature extraction layer
  - Temporal features
  - Relationship features
  - Historical patterns
- [ ] Implement context similarity search
  - Vector index (FAISS/Pinecone)
  - Efficient similarity queries
- [ ] Action item detection model
  - Train or fine-tune NER model
  - Pattern-based extraction

**Deliverables:**
- ML pipeline processing all events
- Embeddings generated and indexed
- Action items detected from content

### Week 6: Context Aggregation & Intelligence
**Goals:** Build intelligent context system

- [ ] Cross-source entity resolution
  - Deduplicate entities across sources
  - Build entity knowledge graph
- [ ] Context window generation
  - Temporal context windows
  - Related event clustering
- [ ] Priority scoring model v2
  - Multi-factor feature engineering
  - Train on user feedback
- [ ] Context-aware summaries
  - Generate summaries of context windows
  - Highlight key action items

**Deliverables:**
- Unified context views across sources
- Improved priority scoring
- Automated context summaries

### Week 7: Proactive Intelligence
**Goals:** Build proactive suggestion system

- [ ] Implement notification engine
  - Priority thresholds
  - Delivery timing optimization
  - Multi-channel delivery
- [ ] Proactive suggestion generation
  - "You have a meeting in 30 min with X, here's the context"
  - "These 3 emails need responses today"
  - "Your task deadline is approaching"
- [ ] Intelligent scheduling assistant
  - Conflict detection
  - Optimal time suggestions
- [ ] Pre-meeting briefs
  - Attendee context
  - Related emails/tasks
  - Suggested talking points

**Deliverables:**
- Proactive notification system live
- Context-aware suggestions working
- Meeting prep automation

### Week 8: Polish, Testing & Documentation
**Goals:** Production readiness

- [ ] Comprehensive testing
  - Unit tests for all components
  - Integration tests for pipelines
  - Load testing for scalability
- [ ] Performance optimization
  - Query optimization
  - Caching strategy
  - Batch processing tuning
- [ ] User feedback collection system
  - Explicit feedback (thumbs up/down)
  - Implicit feedback (dismissals, opens)
  - Feedback incorporation into models
- [ ] Documentation
  - API documentation
  - Architecture documentation
  - User guides
- [ ] Monitoring & alerting
  - Pipeline health dashboards
  - Error rate monitoring
  - Performance metrics

**Deliverables:**
- Production-ready system
- Comprehensive test coverage
- Complete documentation

### Technical Metrics
- **Ingestion latency**: < 5 minutes for all sources
- **Priority score accuracy**: 80%+ based on user feedback
- **Context similarity precision**: Top-5 recall > 70%
- **System uptime**: 99.5%+
- **API response time**: p95 < 500ms

### User Impact Metrics
- **Time saved**: 30+ minutes per day
- **Missed items**: Reduce by 80%
- **Proactive action rate**: 60% of suggestions acted upon
- **User satisfaction**: 4+ stars out of 5

## 🔐 Privacy & Security Considerations

- [ ] End-to-end encryption for sensitive data
- [ ] Granular permission controls per data source
- [ ] Local-first option for privacy-sensitive users
- [ ] Data retention policies
- [ ] GDPR/CCPA compliance
- [ ] Audit logging
- [ ] Secure credential storage (OAuth tokens)

## 🚀 Future Enhancements

- **Collaborative context**: Share context with team members
- **Voice interface**: Verbal context queries and updates
- **Mobile optimization**: On-the-go context access
- **Predictive analytics**: Forecast upcoming busy periods
- **Auto-response suggestions**: Draft replies based on context
- **Meeting notes integration**: Connect to Zoom/Meet transcripts
- **Smart delegation**: Suggest task reassignments

## 📚 Technical Stack

- **Backend**: Python 3.11+, FastAPI, SQLAlchemy
- **ML/AI**: PyTorch, HuggingFace Transformers, sentence-transformers
- **Vector DB**: FAISS (local) or Pinecone (cloud)
- **Database**: PostgreSQL with pgvector extension
- **Task Queue**: Celery + Redis
- **Monitoring**: Prometheus + Grafana
- **Deployment**: Docker, Kubernetes (optional)

### Dependencies
- Email connector (existing)
- OAuth integrations for calendar/task APIs
- ML models (open source)
- Vector search infrastructure

### Risks
- **API rate limits**: Mitigate with caching and smart polling
- **Data quality**: Handle noisy/incomplete data gracefully
- **Cold start**: Limited training data initially
- **Integration complexity**: Different APIs, auth flows
- **Performance**: Large-scale vector search optimization needed

## 📖 References & Resources

- [Sentence Transformers Documentation](https://www.sbert.net/)
- [FAISS: Fast Similarity Search](https://github.com/facebookresearch/faiss)
- [CalDAV Protocol Specification](https://tools.ietf.org/html/rfc4791)
- [Building ML Systems (Book)](https://www.oreilly.com/library/view/designing-machine-learning/9781098107956/)
- [Feature Store Concepts](https://www.featurestore.org/)

---

**Estimated Effort**: 6-8 weeks (1 developer full-time)
**Priority**: High
**Complexity**: High
